# Load the files

In [1]:
from pathlib import Path

import pandas as pd

DATA = Path("../data/raw")

raw = pd.read_parquet(DATA / "prices_se3_raw.parquet")
hourly = pd.read_parquet(DATA / "prices_se3_hourly.parquet")
weather = pd.read_parquet(DATA / "weather_stockholm_hourly.parquet")

for name, df in [("raw prices", raw), ("hourly prices", hourly), ("weather", weather)]:
    print(name, df.shape, "missing values:", df.isna().sum().sum())

raw prices (60311, 5) missing values: 0
hourly prices (34247, 4) missing values: 0
weather (34224, 6) missing values: 0


# Look at each dataset

In [2]:
raw.head(10)

,SEK_per_kWh,EUR_per_kWh,EXR,time_start,time_end
0,0.37995,0.03491,10.883706,2022-10-31 23:00:00+00:00,2022-11-01 00:00:00+00:00
1,0.37995,0.03491,10.883706,2022-11-01 00:00:00+00:00,2022-11-01 01:00:00+00:00
2,0.38430,0.03531,10.883706,2022-11-01 01:00:00+00:00,2022-11-01 02:00:00+00:00
3,0.39301,0.03611,10.883706,2022-11-01 02:00:00+00:00,2022-11-01 03:00:00+00:00
4,0.41173,0.03783,10.883706,2022-11-01 03:00:00+00:00,2022-11-01 04:00:00+00:00
5,0.53134,0.04882,10.883706,2022-11-01 04:00:00+00:00,2022-11-01 05:00:00+00:00
6,0.65248,0.05995,10.883706,2022-11-01 05:00:00+00:00,2022-11-01 06:00:00+00:00
7,0.84882,0.07799,10.883706,2022-11-01 06:00:00+00:00,2022-11-01 07:00:00+00:00
8,0.91858,0.08440,10.883706,2022-11-01 07:00:00+00:00,2022-11-01 08:00:00+00:00
9,0.73628,0.06765,10.883706,2022-11-01 08:00:00+00:00,2022-11-01 09:00:00+00:00


In [3]:
raw.tail(10)

,SEK_per_kWh,EUR_per_kWh,EXR,time_start,time_end
60301,0.94081,0.08328,11.296996,2026-09-27 19:30:00+00:00,2026-09-27 19:45:00+00:00
60302,0.93686,0.08293,11.296996,2026-09-27 19:45:00+00:00,2026-09-27 20:00:00+00:00
60303,1.03797,0.09188,11.296996,2026-09-27 20:00:00+00:00,2026-09-27 20:15:00+00:00
60304,0.93630,0.08288,11.296996,2026-09-27 20:15:00+00:00,2026-09-27 20:30:00+00:00
60305,0.89393,0.07913,11.296996,2026-09-27 20:30:00+00:00,2026-09-27 20:45:00+00:00
60306,0.71058,0.06290,11.296996,2026-09-27 20:45:00+00:00,2026-09-27 21:00:00+00:00
60307,0.73894,0.06541,11.296996,2026-09-27 21:00:00+00:00,2026-09-27 21:15:00+00:00
60308,0.57536,0.05093,11.296996,2026-09-27 21:15:00+00:00,2026-09-27 21:30:00+00:00
60309,0.49537,0.04385,11.296996,2026-09-27 21:30:00+00:00,2026-09-27 21:45:00+00:00
60310,0.29734,0.02632,11.296996,2026-09-27 21:45:00+00:00,2026-09-27 22:00:00+00:00


In [4]:
hourly.describe()

,price_sek_kwh,price_eur_kwh
count,34247.000000,34247.000000
mean,0.617919,0.055437
std,0.665342,0.060378
min,-0.691120,-0.060040
25%,0.192565,0.016970
50%,0.439000,0.039210
75%,0.871517,0.078300
max,8.157090,0.707520


In [5]:
hourly.tail(10)

,time_utc,price_sek_kwh,price_eur_kwh,time_local
34237,2026-09-27 12:00:00+00:00,0.219558,0.019435,2026-09-27 14:00:00+02:00
34238,2026-09-27 13:00:00+00:00,0.233960,0.020710,2026-09-27 15:00:00+02:00
34239,2026-09-27 14:00:00+00:00,0.379408,0.033585,2026-09-27 16:00:00+02:00
34240,2026-09-27 15:00:00+00:00,0.926750,0.082035,2026-09-27 17:00:00+02:00
34241,2026-09-27 16:00:00+00:00,1.270320,0.112448,2026-09-27 18:00:00+02:00
34242,2026-09-27 17:00:00+00:00,1.240100,0.109772,2026-09-27 19:00:00+02:00
34243,2026-09-27 18:00:00+00:00,1.121200,0.099248,2026-09-27 20:00:00+02:00
34244,2026-09-27 19:00:00+00:00,1.006760,0.089118,2026-09-27 21:00:00+02:00
34245,2026-09-27 20:00:00+00:00,0.894695,0.079198,2026-09-27 22:00:00+02:00
34246,2026-09-27 21:00:00+00:00,0.526752,0.046628,2026-09-27 23:00:00+02:00


In [6]:
weather.head(10)

,time_utc,temperature_2m,wind_speed_10m,precipitation,cloud_cover,shortwave_radiation
0,2022-11-01 00:00:00+00:00,9.2,8.8,0.0,100,0.0
1,2022-11-01 01:00:00+00:00,8.7,8.6,0.0,100,0.0
2,2022-11-01 02:00:00+00:00,8.6,8.1,0.0,100,0.0
3,2022-11-01 03:00:00+00:00,8.6,8.5,0.0,100,0.0
4,2022-11-01 04:00:00+00:00,8.4,8.3,0.0,100,0.0
5,2022-11-01 05:00:00+00:00,8.3,7.2,0.0,100,0.0
6,2022-11-01 06:00:00+00:00,8.1,7.2,0.0,100,0.0
7,2022-11-01 07:00:00+00:00,8.2,8.4,0.0,100,5.0
8,2022-11-01 08:00:00+00:00,8.2,8.6,0.0,100,29.0
9,2022-11-01 09:00:00+00:00,8.4,7.2,0.0,100,53.0


In [7]:
weather.describe()

,temperature_2m,wind_speed_10m,precipitation,cloud_cover,shortwave_radiation
count,34224.000000,34224.000000,34224.000000,34224.000000,34224.000000
mean,7.743201,11.877007,0.072362,66.778021,121.566503
std,8.430369,5.403685,0.319330,40.850977,194.119623
min,-18.200000,0.000000,0.000000,0.000000,0.000000
25%,1.300000,8.000000,0.000000,22.000000,0.000000
50%,7.300000,11.300000,0.000000,97.000000,5.000000
75%,14.800000,15.200000,0.000000,100.000000,174.000000
max,30.000000,42.000000,11.900000,100.000000,834.000000


# Check the interval lengths

In [8]:
raw["minutes"] = (raw["time_end"] - raw["time_start"]).dt.total_seconds() / 60
raw["minutes"].value_counts()

minutes
15.0     34751
60.0     25557
120.0        2
75.0         1
Name: count, dtype: int64

# Show the 3 odd rows

In [9]:
odd = raw[~raw["minutes"].isin([15, 60])].copy()
odd["start_local"] = odd["time_start"].dt.tz_convert("Europe/Stockholm")
odd["end_local"] = odd["time_end"].dt.tz_convert("Europe/Stockholm")
odd[["start_local", "end_local", "SEK_per_kWh", "minutes"]]

,start_local,end_local,SEK_per_kWh,minutes
8689,2023-10-29 02:00:00+02:00,2023-10-29 03:00:00+01:00,0.29933,120.0
17425,2024-10-27 02:00:00+02:00,2024-10-27 03:00:00+01:00,-0.00057,120.0
27970,2025-10-26 02:45:00+02:00,2025-10-26 03:00:00+01:00,0.02627,75.0


# Prove the hourly data is not affected

In [10]:
print("duplicate start times:", raw["time_start"].duplicated().sum())
print(hourly["time_utc"].diff().value_counts())

full = pd.date_range(weather["time_utc"].min(), weather["time_utc"].max(), freq="h")
print("missing weather hours:", len(full) - len(weather))

duplicate start times: 0
time_utc
0 days 01:00:00    34246
Name: count, dtype: int64
missing weather hours: 0


## Findings

I checked the three data files before using them for modelling.

**Size and completeness**
- Raw prices: 60,311 rows. Hourly prices: 34,247 rows. Weather: 34,224 rows.
- None of the files have missing values.
- The hourly prices have no gaps. Every row is exactly one hour after the one before.
- The weather data also has no missing hours. It ends on 26 September because the archive is a few days behind.

**Change in price resolution**
- Before 1 October 2025 the prices are hourly. After that they come in 15-minute steps.
- This is why the raw file has more rows than the hourly file.
- I averaged the 15-minute prices into hourly prices so the whole dataset has the same format.

**Problem in the source data**
- Three rows in the raw file have the wrong end time.
- They are on 2023-10-29, 2024-10-27 and 2025-10-26. These are the nights when Sweden goes back to winter time.
- On those nights the clock shows 02:00 twice. The source gives the wrong end time for one of these hours, so it looks like the interval is 75 or 120 minutes long.
- The start times are correct, and there are no duplicate start times.

**Why this does not affect the project**
- My code only uses the start time of each price. The end time is never used.
- The check on the hourly file shows every hour follows the previous one exactly, so no hours are lost or counted twice.

**Weather values**
- Temperature is between -18.2 and 30.0 °C, wind speed up to 42 km/h, and solar radiation from 0 to 834 W/m². All of these look normal for Stockholm.